File: tests/test-keystore.js
#!/usr/bin/env node
'use strict';

const fs = require('fs').promises;
const path = require('path');
const { createKeyStore } = require('../lib/keystores');
const cryptoUtils = require('../lib/crypto-utils');

// Test utilities
let testsPassed = 0;
let testsFailed = 0;

function assert(condition, message) {
  if (!condition) {
    console.error(`✗ ${message}`);
    testsFailed++;
    throw new Error(`Assertion failed: ${message}`);
  }
  testsPassed++;
}

async function cleanup(testDir) {
  try {
    await fs.rm(testDir, { recursive: true, force: true });
  } catch (error) {
    // Ignore cleanup errors
  }
}

// Test FileSystemKeyStore
async function testFileSystemKeyStore() {
  console.log('\n📁 Testing FileSystemKeyStore...');
  const testDir = path.join(__dirname, '.test-keystore');
  
  try {
    await cleanup(testDir);
    
    const keystore = createKeyStore('filesystem', {
      keyStorePath: testDir
    });
    
    await keystore.initialize();
    
    // Test save and load
    const keyPair = await cryptoUtils.generateServerKeyPair();
    const keyData = {
      kid: keyPair.kid,
      publicKey: keyPair.publicKey,
      privateKey: keyPair.privateKey,
      createdAt: Date.now(),
      active: true
    };
    
    await keystore.save(keyData.kid, keyData);
    console.log(`  ✓ Saved key ${keyData.kid}`);
    
    const loaded = await keystore.load(keyData.kid);
    assert(loaded.kid === keyData.kid, 'Loaded key has correct kid');
    assert(loaded.publicKey === keyData.publicKey, 'Loaded key has correct publicKey');
    assert(loaded.privateKey === keyData.privateKey, 'Loaded key has correct privateKey');
    console.log(`  ✓ Loaded key ${keyData.kid}`);
    
    // Test exists
    const exists = await keystore.exists(keyData.kid);
    assert(exists === true, 'Key exists check returns true');
    console.log(`  ✓ Exists check works`);
    
    // Test list
    const list = await keystore.list();
    assert(list.length === 1, 'List returns one key');
    assert(list[0] === keyData.kid, 'List contains correct kid');
    console.log(`  ✓ List returns correct keys`);
    
    // Test loadAll
    const allKeys = await keystore.loadAll();
    assert(allKeys.length === 1, 'LoadAll returns one key');
    assert(allKeys[0].kid === keyData.kid, 'LoadAll contains correct key');
    console.log(`  ✓ LoadAll works`);
    
    // Test delete
    await keystore.delete(keyData.kid);
    const existsAfterDelete = await keystore.exists(keyData.kid);
    assert(existsAfterDelete === false, 'Key deleted successfully');
    console.log(`  ✓ Delete works`);
    
    // Test load non-existent key
    const nonExistent = await keystore.load('non-existent-kid');
    assert(nonExistent === null, 'Load returns null for non-existent key');
    console.log(`  ✓ Load non-existent returns null`);
    
    await keystore.close();
    console.log('✅ FileSystemKeyStore tests passed');
    
  } finally {
    await cleanup(testDir);
  }
}

// Test MemoryKeyStore
async function testMemoryKeyStore() {
  console.log('\n💾 Testing MemoryKeyStore...');
  
  const keystore = createKeyStore('memory');
  await keystore.initialize();
  
  // Test save and load
  const keyPair = await cryptoUtils.generateServerKeyPair();
  const keyData = {
    kid: keyPair.kid,
    publicKey: keyPair.publicKey,
    privateKey: keyPair.privateKey,
    createdAt: Date.now(),
    active: true
  };
  
  await keystore.save(keyData.kid, keyData);
  console.log(`  ✓ Saved key ${keyData.kid}`);
  
  const loaded = await keystore.load(keyData.kid);
  assert(loaded.kid === keyData.kid, 'Loaded key has correct kid');
  console.log(`  ✓ Loaded key ${keyData.kid}`);
  
  // Test getStats
  const stats = keystore.getStats();
  assert(stats.totalKeys === 1, 'Stats shows 1 key');
  console.log(`  ✓ Stats: ${JSON.stringify(stats)}`);
  
  // Test delete
  await keystore.delete(keyData.kid);
  const existsAfterDelete = await keystore.exists(keyData.kid);
  assert(existsAfterDelete === false, 'Key deleted successfully');
  console.log(`  ✓ Delete works`);
  
  await keystore.close();
  console.log('✅ MemoryKeyStore tests passed');
}

// Test keystore factory
async function testKeystoreFactory() {
  console.log('\n🏭 Testing Keystore Factory...');
  
  // Test filesystem creation
  const fsStore = createKeyStore('filesystem', { keyStorePath: '.test-factory-fs' });
  assert(fsStore.constructor.name === 'FileSystemKeyStore', 'Factory creates FileSystemKeyStore');
  console.log('  ✓ Factory creates FileSystemKeyStore');
  
  // Test memory creation
  const memStore = createKeyStore('memory');
  assert(memStore.constructor.name === 'MemoryKeyStore', 'Factory creates MemoryKeyStore');
  console.log('  ✓ Factory creates MemoryKeyStore');
  
  // Test mongodb creation (without actual client - just verify class instantiation)
  try {
    createKeyStore('mongodb', { connectionString: 'mongodb://localhost:27017' });
    assert(false, 'Factory should throw without MongoDB client');
  } catch (error) {
    assert(error.message.includes('MongoDB client is required'), 'Factory requires MongoDB client');
    console.log('  ✓ Factory validates MongoDB client requirement');
  }
  
  // Test invalid type
  try {
    createKeyStore('invalid-type');
    assert(false, 'Factory should throw for invalid type');
  } catch (error) {
    assert(error.message.includes('Unknown keystore type'), 'Factory throws for invalid type');
    console.log('  ✓ Factory throws for invalid type');
  }
  
  console.log('✅ Keystore Factory tests passed');
  
  // Cleanup
  await cleanup('.test-factory-fs');
}

// Test EncryptedServer with keystore persistence
async function testServerPersistence() {
  console.log('\n🔐 Testing EncryptedServer Key Persistence...');
  const testDir = path.join(__dirname, '.test-server-keystore');
  
  try {
    await cleanup(testDir);
    
    const EncryptedServer = require('../lib/encrypted-server');
    
    // Create server with filesystem keystore
    const server1 = new EncryptedServer({
      port: 3001,
      saveDirectory: path.join(testDir, 'uploads'),
      keystoreType: 'filesystem',
      keystoreOptions: {
        keyStorePath: path.join(testDir, 'keys')
      }
    });
    
    await server1.listen(3001);
    const kid1 = server1.activeKeyId;
    console.log(`  ✓ Server 1 started with key ${kid1}`);
    
    // Close and restart
    server1.close();
    console.log(`  ✓ Server 1 closed`);
    
    // Create new server instance (simulating restart)
    const server2 = new EncryptedServer({
      port: 3002,
      saveDirectory: path.join(testDir, 'uploads'),
      keystoreType: 'filesystem',
      keystoreOptions: {
        keyStorePath: path.join(testDir, 'keys')
      }
    });
    
    await server2.listen(3002);
    const kid2 = server2.activeKeyId;
    console.log(`  ✓ Server 2 started with key ${kid2}`);
    
    // Verify key persistence
    assert(kid2 === kid1, 'Server loaded same key after restart');
    console.log(`  ✓ Key persisted across restart`);
    
    // Verify old key can still decrypt
    const privateKey = server2.getPrivateKey(kid1);
    assert(privateKey !== null, 'Old key still available for decryption');
    console.log(`  ✓ Old key still available for decryption`);
    
    server2.close();
    console.log('✅ Server persistence tests passed');
    
  } finally {
    await cleanup(testDir);
  }
}

// Main test runner
async function runTests() {
  console.log('🧪 Keystore Integration Tests\n');
  console.log('='.repeat(50));
  
  try {
    await testFileSystemKeyStore();
    await testMemoryKeyStore();
    await testKeystoreFactory();
    await testServerPersistence();
    
    console.log('\n' + '='.repeat(50));
    console.log(`\n✅ All tests passed! (${testsPassed} assertions)`);
    console.log(`   Tests passed: ${testsPassed}`);
    console.log(`   Tests failed: ${testsFailed}`);
    process.exit(0);
  } catch (error) {
    console.error('\n' + '='.repeat(50));
    console.error(`\n✗ Tests failed: ${error.message}`);
    console.error(`   Tests passed: ${testsPassed}`);
    console.error(`   Tests failed: ${testsFailed}`);
    process.exit(1);
  }
}

// Run if called directly
if (require.main === module) {
  runTests().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

module.exports = { runTests };

File: python/tests/test_keystore.py
"""
Comprehensive tests for keystore implementations
"""

import os
import json
import stat
import pytest
import tempfile
import shutil
from pathlib import Path

from indexedcp.keystores import (
    BaseKeyStore,
    FileSystemKeyStore,
    create_keystore
)
from indexedcp import CryptoUtils


class TestBaseKeyStore:
    """Test BaseKeyStore abstract class"""
    
    def test_cannot_instantiate_directly(self):
        """Test that BaseKeyStore cannot be instantiated directly"""
        with pytest.raises(TypeError):
            BaseKeyStore()
    
    def test_abstract_methods_must_be_implemented(self):
        """Test that subclasses must implement abstract methods"""
        
        class IncompleteKeyStore(BaseKeyStore):
            pass
        
        with pytest.raises(TypeError):
            IncompleteKeyStore()


class TestFileSystemKeyStore:
    """Test FileSystemKeyStore implementation"""
    
    @pytest.fixture
    async def temp_keystore_dir(self):
        """Create temporary directory for keystore tests"""
        temp_dir = tempfile.mkdtemp(prefix='test_keystore_')
        yield temp_dir
        # Cleanup
        shutil.rmtree(temp_dir, ignore_errors=True)
    
    @pytest.fixture
    async def keystore(self, temp_keystore_dir):
        """Create filesystem keystore instance"""
        ks = FileSystemKeyStore({'key_store_path': temp_keystore_dir})
        await ks.initialize()
        yield ks
        await ks.close()
    
    @pytest.fixture
    def sample_key_data(self):
        """Generate sample key data for testing"""
        crypto = CryptoUtils()
        key_pair = crypto.generate_server_key_pair()
        return {
            'kid': key_pair['kid'],
            'publicKey': key_pair['publicKey'],
            'privateKey': key_pair['privateKey'],
            'createdAt': 1699999999999,
            'active': True
        }
    
    @pytest.mark.asyncio
    async def test_initialize_creates_directory(self, temp_keystore_dir):
        """Test that initialize creates the keystore directory"""
        keystore = FileSystemKeyStore({'key_store_path': temp_keystore_dir})
        await keystore.initialize()
        
        assert Path(temp_keystore_dir).exists()
        assert Path(temp_keystore_dir).is_dir()
        
        # Check directory permissions (0700)
        dir_stat = os.stat(temp_keystore_dir)
        dir_mode = stat.S_IMODE(dir_stat.st_mode)
        assert dir_mode == stat.S_IRWXU  # 0700
    
    @pytest.mark.asyncio
    async def test_save_and_load(self, keystore, sample_key_data):
        """Test saving and loading a key"""
        # Save key
        await keystore.save(sample_key_data['kid'], sample_key_data)
        
        # Load key
        loaded = await keystore.load(sample_key_data['kid'])
        
        assert loaded is not None
        assert loaded['kid'] == sample_key_data['kid']
        assert loaded['publicKey'] == sample_key_data['publicKey']
        assert loaded['privateKey'] == sample_key_data['privateKey']
        assert loaded['createdAt'] == sample_key_data['createdAt']
        assert loaded['active'] == sample_key_data['active']
    
    @pytest.mark.asyncio
    async def test_save_sets_file_permissions(self, keystore, sample_key_data, temp_keystore_dir):
        """Test that saved key files have correct permissions (0600)"""
        await keystore.save(sample_key_data['kid'], sample_key_data)
        
        key_file = Path(temp_keystore_dir) / f"{sample_key_data['kid']}.json"
        file_stat = os.stat(key_file)
        file_mode = stat.S_IMODE(file_stat.st_mode)
        
        # Check file permissions (0600 - owner read/write only)
        assert file_mode == (stat.S_IRUSR | stat.S_IWUSR)  # 0600
    
    @pytest.mark.asyncio
    async def test_load_nonexistent_returns_none(self, keystore):
        """Test that loading non-existent key returns None"""
        loaded = await keystore.load('non-existent-kid')
        assert loaded is None
    
    @pytest.mark.asyncio
    async def test_exists(self, keystore, sample_key_data):
        """Test key existence check"""
        # Key doesn't exist yet
        assert await keystore.exists(sample_key_data['kid']) is False
        
        # Save key
        await keystore.save(sample_key_data['kid'], sample_key_data)
        
        # Key now exists
        assert await keystore.exists(sample_key_data['kid']) is True
    
    @pytest.mark.asyncio
    async def test_delete(self, keystore, sample_key_data):
        """Test deleting a key"""
        # Save key
        await keystore.save(sample_key_data['kid'], sample_key_data)
        assert await keystore.exists(sample_key_data['kid']) is True
        
        # Delete key
        result = await keystore.delete(sample_key_data['kid'])
        assert result is True
        
        # Key no longer exists
        assert await keystore.exists(sample_key_data['kid']) is False
    
    @pytest.mark.asyncio
    async def test_delete_nonexistent_returns_false(self, keystore):
        """Test that deleting non-existent key returns False"""
        result = await keystore.delete('non-existent-kid')
        assert result is False
    
    @pytest.mark.asyncio
    async def test_list(self, keystore):
        """Test listing key IDs"""
        # Create multiple keys
        crypto = CryptoUtils()
        keys = []
        for _ in range(3):
            kp = crypto.generate_server_key_pair()
            key_data = {
                'kid': kp['kid'],
                'publicKey': kp['publicKey'],
                'privateKey': kp['privateKey'],
                'createdAt': 1699999999999,
                'active': False
            }
            keys.append(key_data)
            await keystore.save(key_data['kid'], key_data)
        
        # List keys
        key_list = await keystore.list()
        
        assert len(key_list) == 3
        for key_data in keys:
            assert key_data['kid'] in key_list
    
    @pytest.mark.asyncio
    async def test_load_all(self, keystore):
        """Test loading all keys"""
        # Create multiple keys
        crypto = CryptoUtils()
        keys = []
        for _ in range(3):
            kp = crypto.generate_server_key_pair()
            key_data = {
                'kid': kp['kid'],
                'publicKey': kp['publicKey'],
                'privateKey': kp['privateKey'],
                'createdAt': 1699999999999,
                'active': False
            }
            keys.append(key_data)
            await keystore.save(key_data['kid'], key_data)
        
        # Load all keys
        all_keys = await keystore.load_all()
        
        assert len(all_keys) == 3
        loaded_kids = {k['kid'] for k in all_keys}
        expected_kids = {k['kid'] for k in keys}
        assert loaded_kids == expected_kids
    
    @pytest.mark.asyncio
    async def test_load_all_empty_directory(self, keystore):
        """Test loading all keys from empty directory"""
        all_keys = await keystore.load_all()
        assert all_keys == []
    
    @pytest.mark.asyncio
    async def test_json_format(self, keystore, sample_key_data, temp_keystore_dir):
        """Test that keys are stored in proper JSON format"""
        await keystore.save(sample_key_data['kid'], sample_key_data)
        
        # Read file directly
        key_file = Path(temp_keystore_dir) / f"{sample_key_data['kid']}.json"
        with open(key_file, 'r') as f:
            data = json.load(f)
        
        assert data == sample_key_data
    
    @pytest.mark.asyncio
    async def test_file_locking_concurrent_writes(self, keystore, sample_key_data):
        """Test that file locking works for concurrent operations"""
        import asyncio
        
        # Attempt concurrent saves (file locking should handle this)
        async def save_operation(kid_suffix):
            key_data = sample_key_data.copy()
            key_data['kid'] = f"{sample_key_data['kid']}_{kid_suffix}"
            await keystore.save(key_data['kid'], key_data)
        
        # Run concurrent saves
        await asyncio.gather(
            save_operation('1'),
            save_operation('2'),
            save_operation('3')
        )
        
        # Verify all keys were saved
        all_keys = await keystore.load_all()
        assert len(all_keys) == 3
    
    @pytest.mark.asyncio
    async def test_persistence_across_instances(self, temp_keystore_dir, sample_key_data):
        """Test that keys persist across keystore instances"""
        # Create first instance and save key
        keystore1 = FileSystemKeyStore({'key_store_path': temp_keystore_dir})
        await keystore1.initialize()
        await keystore1.save(sample_key_data['kid'], sample_key_data)
        await keystore1.close()
        
        # Create second instance and load key
        keystore2 = FileSystemKeyStore({'key_store_path': temp_keystore_dir})
        await keystore2.initialize()
        loaded = await keystore2.load(sample_key_data['kid'])
        await keystore2.close()
        
        assert loaded is not None
        assert loaded['kid'] == sample_key_data['kid']


class TestKeystoreFactory:
    """Test keystore factory function"""
    
    def test_create_filesystem_keystore(self):
        """Test factory creates FileSystemKeyStore"""
        keystore = create_keystore('filesystem', {'key_store_path': './test-keys'})
        assert isinstance(keystore, FileSystemKeyStore)
    
    def test_create_filesystem_keystore_aliases(self):
        """Test factory recognizes filesystem aliases"""
        # Test all aliases
        for alias in ['filesystem', 'file', 'fs']:
            keystore = create_keystore(alias, {'key_store_path': './test-keys'})
            assert isinstance(keystore, FileSystemKeyStore)
    
    def test_create_filesystem_keystore_case_insensitive(self):
        """Test factory is case-insensitive"""
        keystore = create_keystore('FileSystem', {'key_store_path': './test-keys'})
        assert isinstance(keystore, FileSystemKeyStore)
    
    def test_factory_invalid_type_raises_error(self):
        """Test factory raises error for invalid type"""
        with pytest.raises(ValueError) as exc_info:
            create_keystore('invalid-type')
        
        assert 'Unknown keystore type' in str(exc_info.value)
        assert 'invalid-type' in str(exc_info.value)
    
    def test_factory_with_options(self):
        """Test factory passes options to keystore"""
        keystore = create_keystore('filesystem', {
            'key_store_path': './custom-path',
            'log_level': 'DEBUG'
        })
        assert isinstance(keystore, FileSystemKeyStore)
        assert keystore.key_store_path == Path('./custom-path')


class TestKeystoreIntegration:
    """Integration tests for keystore with crypto operations"""
    
    @pytest.fixture
    async def temp_dir(self):
        """Create temporary directory"""
        temp_dir = tempfile.mkdtemp(prefix='test_integration_')
        yield temp_dir
        shutil.rmtree(temp_dir, ignore_errors=True)
    
    @pytest.mark.asyncio
    async def test_full_key_lifecycle(self, temp_dir):
        """Test complete key lifecycle: generate, save, load, use, delete"""
        # Create keystore
        keystore = create_keystore('filesystem', {'key_store_path': temp_dir})
        await keystore.initialize()
        
        # Generate key pair
        crypto = CryptoUtils()
        key_pair = crypto.generate_server_key_pair()
        
        # Save to keystore
        key_data = {
            'kid': key_pair['kid'],
            'publicKey': key_pair['publicKey'],
            'privateKey': key_pair['privateKey'],
            'createdAt': 1699999999999,
            'active': True
        }
        await keystore.save(key_data['kid'], key_data)
        
        # Load from keystore
        loaded = await keystore.load(key_data['kid'])
        assert loaded is not None
        
        # Use the loaded key for encryption/decryption
        session_key = crypto.generate_session_key()
        wrapped = crypto.wrap_session_key(session_key, loaded['publicKey'])
        unwrapped = crypto.unwrap_session_key(wrapped, loaded['privateKey'])
        assert session_key == unwrapped
        
        # Delete key
        await keystore.delete(key_data['kid'])
        assert await keystore.exists(key_data['kid']) is False
        
        await keystore.close()
    
    @pytest.mark.asyncio
    async def test_key_rotation_scenario(self, temp_dir):
        """Test key rotation: new key active, old key retained for decryption"""
        keystore = create_keystore('filesystem', {'key_store_path': temp_dir})
        await keystore.initialize()
        
        crypto = CryptoUtils()
        
        # Generate and save old key (initially active)
        old_kp = crypto.generate_server_key_pair()
        old_key = {
            'kid': old_kp['kid'],
            'publicKey': old_kp['publicKey'],
            'privateKey': old_kp['privateKey'],
            'createdAt': 1699999999999,
            'active': True
        }
        await keystore.save(old_key['kid'], old_key)
        
        # Generate and save new key (now active)
        new_kp = crypto.generate_server_key_pair()
        new_key = {
            'kid': new_kp['kid'],
            'publicKey': new_kp['publicKey'],
            'privateKey': new_kp['privateKey'],
            'createdAt': 1700000000000,
            'active': True
        }
        
        # Deactivate old key
        old_key['active'] = False
        await keystore.save(old_key['kid'], old_key)
        
        # Save new key
        await keystore.save(new_key['kid'], new_key)
        
        # Verify both keys exist
        assert await keystore.exists(old_key['kid']) is True
        assert await keystore.exists(new_key['kid']) is True
        
        # Load both keys
        loaded_old = await keystore.load(old_key['kid'])
        loaded_new = await keystore.load(new_key['kid'])
        
        assert loaded_old['active'] is False
        assert loaded_new['active'] is True
        
        # Old key can still decrypt old data
        session_key = crypto.generate_session_key()
        wrapped_with_old = crypto.wrap_session_key(session_key, loaded_old['publicKey'])
        unwrapped = crypto.unwrap_session_key(wrapped_with_old, loaded_old['privateKey'])
        assert session_key == unwrapped
        
        await keystore.close()
    
    @pytest.mark.asyncio
    async def test_multiple_active_keys(self, temp_dir):
        """Test scenario with multiple keys for multi-server deployment"""
        keystore = create_keystore('filesystem', {'key_store_path': temp_dir})
        await keystore.initialize()
        
        crypto = CryptoUtils()
        
        # Create 3 keys (simulating 3 servers)
        keys = []
        for i in range(3):
            kp = crypto.generate_server_key_pair()
            key_data = {
                'kid': kp['kid'],
                'publicKey': kp['publicKey'],
                'privateKey': kp['privateKey'],
                'createdAt': 1699999999999 + i,
                'active': True
            }
            keys.append(key_data)
            await keystore.save(key_data['kid'], key_data)
        
        # Verify all keys stored
        all_keys = await keystore.load_all()
        assert len(all_keys) == 3
        
        # Each key should work independently
        session_key = crypto.generate_session_key()
        for key_data in keys:
            loaded = await keystore.load(key_data['kid'])
            wrapped = crypto.wrap_session_key(session_key, loaded['publicKey'])
            unwrapped = crypto.unwrap_session_key(wrapped, loaded['privateKey'])
            assert session_key == unwrapped
        
        await keystore.close()
